aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws.s3;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3URI;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.S3ObjectInputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import org.apache.commons.compress.utils.IOUtils;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class S3InputStream extends SeekableInputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(S3InputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final AmazonS3 s3;
+  private final AmazonS3URI location;
+
+  private S3ObjectInputStream stream;
+  private long pos = 0;
+  private long next = 0;
+  private boolean closed = false;
+
+  private int skipSize = 1024 * 1024;
+
+  public S3InputStream(AmazonS3 s3, AmazonS3URI location) {
+    this.s3 = s3;
+    this.location = location;
+
+    createStack = Thread.currentThread().getStackTrace();
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "already closed");
+    Preconditions.checkArgument(newPos >= 0, "position is negative: %s", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
+    next = newPos;
+  }
+
+  @Override
+  public int read() {
+    // This stream is wrapped with BufferedInputStream, so this method should never be called
+    throw new UnsupportedOperationException();
```

Comment:
Why not implement it anyway? All we need to do is delegate to stream, right?

Reply:
One argument would be that if you're hitting this, you're doing something wrong (and risk poor performance). Throwing here is a form of defensive programming.

Reply:
@jacques-n, isn't the underlying stream buffered, though? Why require wrapping this one in a BufferedStream as well?

There are a few places where we use `read()` directly, assuming the data is buffered. Like in the vectorized reader base class: https://github.com/apache/iceberg/blob/master/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/BaseVectorizedParquetValuesReader.java#L147

A one significant speedup we saw when working on vectorization was avoiding code in Parquet that would read into a 4- or 8-byte buffer and then access each byte individual to reconstruct numeric values. Maybe there is a better way that what we're currently doing though?

Reply:
Good questions. I haven't looked at the underlying reader to confirm so it may be a moot point.

Generally, in our other performance work we've tried to move away from InputStream entirely. For example, we expose this as the only input mechanism for streams: 

https://github.com/dremio/dremio-oss/blob/master/common/src/main/java/com/dremio/io/AsyncByteReader.java#L41

Then higher levels should be preallocating space based on say, page size, do a bulk pull from the underlying system (say s3) and then doing all byte reads using unsafe byte reads or similar (as opposed to an inputstream stack).

For reference, the impl for S3 of the above are in these two classes: [Using AWS Sync Client](https://github.com/dremio/dremio-oss/blob/master/plugins/s3/src/main/java/com/dremio/plugins/s3/store/S3AsyncByteReaderUsingSyncClient.java), [Useing AWS Async Client](https://github.com/dremio/dremio-oss/blob/master/plugins/s3/src/main/java/com/dremio/plugins/s3/store/S3AsyncByteReader.java)

Feel free to grab anything that seems useful. Unfortunately, we don't actually use the latter true async implementation because we've found the AWS Java SDk2 async api for S3 to be very buggy (or at least it was when were last testing it). In cases for Azure and similar, we actually chose to use the relatively strong [AsyncHttpClient|https://github.com/AsyncHttpClient/async-http-client] as you can find [here|https://github.com/dremio/dremio-oss/blob/master/plugins/azure/src/main/java/com/dremio/plugins/azure/AzureAsyncReader.java].

Reply:
@jacques-n, we have some similar optimizations as well and I think they would be great follow ons to an initial implementation.  For example, we have a vectored-io implementation for Parquet that will only do bounded range gets from S3, which is more efficient for connections and limits any buffered reads by the sdk past the intended range.

However, for now, I think we should keep this relatively straightforward so we have a working initial point.

I'll definitely take a look at the sync and async clients, those look promising.

Reply:
I'm in agreement on getting a working initial implementation in and then optimizing it later. We use similar optimizations when reading / writing directly to S3 from plain old java programs.

This might somewhat go against the grain of how `FileIO` works, but if there were a filesystem concept like exists in some of the python code, we could make use of the multipart file upload manager and then upload based on the configured block size, which makes reads from S3 (when reading along the same block size) significantly more performant. It also means that all file uploads have a final commit operation (such that they all succeed or don't). I guess with Iceberg that final commit operation isn't really necessary though, as it's ok for tables to have files in them that don't belong as they will be filtered out by the metadata.

But I'm still hugely in favor of making multipart uploads work for us. As well as a more performant way to download only the portion of the file thats needed (e.g. the footer, etc).

Reply:
@danielcweeks, totally agreed on my comments around patterns not being important notes for this patch. I was only meaning to share some of the ways we've moved away from inputstream in our code.

Reply:
@kbendick I think a few other threads on here have touched on some of your points.  We do want to do progressive multipart upload as a follow up.

For the read, I think there a few different paths forward as well.  We've implemented a vectored-io to enable more performant reads, and as @jacques-n mentioned, there are other optimizations we can look into for the read path.  These get a little more complicated and may affect some upstream dependencies (like parquet), so I expect to evolve this initial approach.

Reply:
So, back to @rdblue's original comment, I did a little digging to see what is underlying the stream from the aws sdk and there is buffering happening via a `SessionInputBuffer` on the request.  I'm going to go ahead and implement the read, but I also agree with @jacques-n that it's probably an issue if this ends up being called for performance reasons.